### PR TITLE
fix: fix param grad sync under SP

### PR DESCRIPTION
**Pull request by Chamberlain0w0**
https://gxtctab8no8.feishu.cn/docx/FR88dJOZPoXT1Ox1TH1c2abpnof

### CODE REVIEW

File: infini_train/include/optimizer.h [hunk 1]
```diff
@@ -27,7 +28,11 @@ class Optimizer {
 
     virtual void ZeroGrad(bool set_to_none = true);
 
-    virtual void Step() = 0;
+    virtual void Step() final;
+
+    void set_model_grad_finalizer(ModelGradFinalizer finalizer) { model_grad_finalizer_ = std::move(finalizer); }
+
+    const std::vector<std::shared_ptr<Tensor>> &parameters() const { return params_; }
```
Comment (kilinchange) [inline infini_train/include/optimizer.h]: 实现放 .cc 里。

File: infini_train/include/tensor.h [hunk 2]
```diff
@@ -80,6 +80,9 @@ class Tensor : public std::enable_shared_from_this<Tensor> {
     size_t NumElements() const;
     DataType Dtype() const;
 
+    void set_sequence_parallel(bool enabled) { sequence_parallel_ = enabled; }
+    bool sequence_parallel() const { return sequence_parallel_; }
```
Comment (kilinchange) [inline infini_train/include/tensor.h:84]: 实现放 .cc 里。

File: infini_train/src/nn/modules/normalization.cc
Comment (kilinchange) [inline infini_train/src/nn/modules/normalization.cc:1]: 这里得根据 parallel::global::GetSequenceParallelEnabled() 决定是否 set 吧

File: example/gpt2/main.cc [hunk 3]
```diff
@@ -351,6 +351,7 @@ void Train(const nn::parallel::Rank &rank) {
     } else {
         optimizer = optimizer_creator(named_parameters);
     }
+    optimizer->set_model_grad_finalizer(nn::parallel::FinalizeModelGrads);
```
Comment (kilinchange) [inline example/gpt2/main.cc:354]: FinalizeModelGrads 绑定到 optimizer 上感觉职责不太合适。optimizer 接收到的 grad 应该已经是 finalize 后可用于更新的完整梯度，本身只负责后续的 clip / step 等操作。

另外，将 FinalizeModelGrads 放到 optimizer->Step() 内部也容易影响 clip 执行顺序（理论上应该保证：
FinalizeModelGrads -> clip -> step，因为 clip 需要基于最终同步完成的梯度计算）。

建议目前在 optimizer->Step() 之前显式调用 nn::parallel::FinalizeModelGrads(named_parameters)；后续如果抽象统一的 training/schedule 入口，再将这部分逻辑收进去。Megatron 也是在 forward/backward schedule 结束后调用 finalize_model_grads，而不是绑定到 optimizer：

（no pipeline schedule 情况，带 pipeline 时类似）https://github.com/NVIDIA/Megatron-LM/blob/0ac6ffd3859fef41bbfcd92a67d56861a79d4b34/megatron/core/pipeline_parallel/schedules.py#L866

**Review by kilinchange — CHANGES_REQUESTED** [review]

File: infini_train/include/tensor.h (hunk 2 above)
Comment (Chamberlain0w0) [reply infini_train/include/tensor.h]: 改了

File: infini_train/src/nn/modules/normalization.cc
Comment (Chamberlain0w0) [reply infini_train/src/nn/modules/normalization.cc:1]: 改了

File: example/gpt2/main.cc (hunk 3 above)
Comment (Chamberlain0w0) [reply example/gpt2/main.cc]: 改了一下。现在是 training loop 里面需要显式调用 nn::parallel::FinalizeModelGrads，但是传入的是 model chunks vector。在 backward 完成、optimizer 更新前，统一调用 FinalizeModelGrads(model_chunks)：先完成 DP 同步，再执行 SP 梯度的 TP SUM。

另外对应做的修改：
1. 原位于 DistOpt 里面的 StartGradSync() 和 FinishGradSync() 交由 DDP model 管理；同时删除 StartGradSync()，因为 DDP 里面实际同步开启时机相对固定，简化起见目前不需要额外留这个用户接口；
2. 未做引入 main_grad 基建的设计，而是让 bucket group 新增 LocalGradShard 映射表：DistOpt 在原有 BuildShardParamsAndBindGrads() 中登记原参数与本地梯度 view，finalizer 据此处理 ZeRO-1/2，尤其解决 ZeRO-2 原参数 grad() 为空时的梯度访问问题。只保存引用，不新增显存 buffer。

    > 问题简述：
    > 我们现在实现没有一个统一的 main_grad 的全局 buffer，并且 zero2 里面切梯度，只在反向过程中会算临时 full grad，一方面不会把这个 full grad 写入 tensor.grad，RS 之后也不会把 grad shard 写到 tensor.grad 里面（不符合 tensor.shape == grad.shape 的约束）。之后 finalize 统一入口没法通过 param 获取对应的 param.grad，可能需要额外的维护一个 map

File: infini_train/src/tensor.cc [hunk 4]
```diff
@@ -170,6 +175,7 @@ Tensor Tensor::To(Device device) {
     }
 
     new_tensor.requires_grad_ = requires_grad_;
+    new_tensor.sequence_parallel_ = sequence_parallel_;
```
Comment (kilinchange) [inline infini_train/src/tensor.cc]: 这里 Tensor::To 函数在 device == buffer_device 的快速路径，requires_grad_ 和 seqence_parallel_ 属性都被丢弃了，需要修复一下。

File: example/llama3/main.cc [hunk 5]
```diff
@@ -503,6 +502,7 @@ void Train(const nn::parallel::Rank &rank) {
                 LOG(INFO) << "Rank " << rank.GlobalRank() << ": finish backward";
             }
 
+            nn::parallel::FinalizeModelGrads({model});
```
Comment (kilinchange) [inline example/llama3/main.cc]: mixtral 也加一下吧

**Review by kilinchange — CHANGES_REQUESTED** [review]
